### DISCUSSION TITLE
Composer interest in PHP 8.1 accelerated include resolving

### DISCUSSION THREAD

**Discussion by markrandall** [composer/composer#9906] [General]
Greetings,

I am currently working on a PHP RFC for an accelerated autoload classmap, this RFC adds a new mechanism called autoload_set_classmap() which accepts an associative array of classname => file paths, and acts as an internal first-stage autoloader.

The change offers a 5% improvement in performance vs hitting composer's existing userland function and performing the lookup there (testing exclusively the autoload mechanism). As composer is the de-facto autoloader, the general consensus is that it would only be worth adding if Composer itself could take advantage of it.

The necessary change would be to call autoload_set_classmap as part of the existing autoloader registration, feeding in the existing static autoload classmap, with one exception, class names must be stored lowercase (as per internal class naming). This would be something that would be necessary to do as part of the dump autoload step.

Therefore, I ask would composer have an interest in using such a function should it be accepted for 8.1.

**Comment by stof** [comment]
To me, this makes perfect sense.

**Comment by Seldaek** [comment]
I have emailed you twice on the internals list about this:

https://externals.io/message/113545#113984
and previously https://externals.io/message/113576#113562

I still haven't seen an answer from you about the concerns I raised regarding having multiple autoloaders in one process.

**Reply by markrandall** [reply]
The classmap mechanism is intended for accelerating a singular first-step lookup table that has been pre-built to be as complete as possible, thus only a single map is supported to offer the fastest lookups. 

I have not seen any examples in the wild of an autoloader prepending itself for the purposes of overriding what would be included in the composer classmap, do you have examples of where this is the case that I could look into? 

If more complex behaviour is required where overrides are required, then existing autoload functions are already in place and should be used instead without a classmap. This is because even if a library chose to prepend its own autoloader, it would need to know to prepend its own classmap too, as classmaps would exist in their own priority list but would still always be checked first, to do otherwise would require changing spl_autoload_register to accept both functions and classmaps at once, preserving them in order and significantly confusing the the related functions such as spl_autoload_functions and unregister.

In terms of it coming first, I don't consider it as too-significant a change as PHP does not give any existing guarantees that a manually prepended autoloader will stay at the front as any other part of the code could always prepend in front of it. By contrast, autoload_set_classmap does guarantee it runs first.

The RFC was already updated to use require, rather than require_once. An error if the file could not be found is appropriate as it suggests something is wrong with the data provided to the map.

**Reply by stof** [reply]
@markrandall it is quite common to have multiple composer autoloader getting registered in the same PHP process. A common case would be a phar (that uses a composer-generated autoloader internally) that also needs to load the code of the project (think phpunit for instance). Or a globally installed tool (with composer) that also loads code from the project.

Manually merging classmaps before registering them has a huge drawback: it means that the array cannot stay in shared memory and has to be copied in the process (to be able to modify it).
A full composer classmap can be massive. In the past (see https://github.com/composer/composer/pull/5174), Composer achieved a 10% perf increase (and a huge memory decrease) by ensuring that the classmap can stay in the opcache shared memory.

**Reply by markrandall** [reply]
> A full composer classmap can be massive. In the past (see #5174), Composer achieved a 10% perf increase (and a huge memory decrease) by ensuring that the classmap can stay in the opcache shared memory.

The autoload_set_classmap adds a reference to the existing array, rather than copying it in place or modifying it. It will therefore use the same copy of the array from opcache. 

I am not attempting to argue that there is no need for multiple autoloaders, but based purely on my own experience, well in excess of 99% of autoloaded classes are able to be hit by a build-time classmap generated from files in userland and the vendor directory, and additional autoloaders pretty much exclusively deal with following their own non-standard naming convention for their own library.

If this experience is wrong, I am happy to be corrected.

But assuming it is correct, this RFC seeks to transparently increase the performance of the 99%, allowing the remaining 1% to continue on to the existing autoloader functions.

**Reply by stof** [reply]
@markrandall the issue of the copy is not `autoload_set_classmap`. It is the suggested way to support multiple autoloaders by fetching the existing classmap and merge it. That merging would trigger a copy.

**Reply by stof** [reply]
and the discussion about the case of multiple composer generated autoloaders being registered in the process is not about handling the case of non standard autoloading. The composer generated autoloader cannot assume that another one as not been loaded already. If only one of them can register its classmap, this likely means that composer will not be able to safely use that feature when dumping the autoloader (which then forbids you from improving things for the 99%)

**Reply by markrandall** [reply]
>  (which then forbids you from improving things for the 99%)

I believe that would be an incredibly niche case, I must stress that autoload_set_classmap is designed for global autoloading frameworks processed at build-time, not individual libraries. 

IMO there shouldn't be any merging or modifying the classmap, ever.

**Reply by stof** [reply]
@markrandall I'm **not** talking about individual libraries. I'm talking about the fact that it is perfectly valid for multiple autoloaders generated by composer to end up being registered in the same PHP process (and not by accident. Parts of it have been specifically improved over time to ensure it works fine).
So if `autoload_set_classmap` is incompatible with such setup, it means Composer will need to avoid using it for its generated autoloader, which in turns means that none of the modern PHP frameworks will use the feature (as they all delegate autoloading to composer).

**Reply by markrandall** [reply]
I feel I am struggling to communicate the benefits, as you are seemingly thinking of autoload_set_classmap along the same lines as spl_autoload_register, this is not the case. 

I have attempted to express that this feature is not designed for multiple autoloaders, because it is designed to reflect the modern paradigm of application-wide pre-generated classmaps that were simply not a common thing when the original mechanism was added (spl_autoload_register was added in PHP 5.1, 2006  whereas composer reached version 1.0.0 in 2016).

That is no longer the case. Thanks primarily to Composer, the majority of modern applications are now built with classmaps generated at build time, however internals has not reflected that paradigm shift by providing more performant methods to accelerate those common operations, that is what this RFC seeks to address.

It is the very fact that all other frameworks delegate their autoloading to composer that makes it an ideal opportunity for a transparent performance increase. Almost all modern code invokes the composer autoloader as one of its first operations for each request, this  would allow composer to provide the engine with its pre-built class map.

Please correct me if I am wrong, but your objection seems to be that because it cannot be used for the < 1% case, it's not worth using for the > 99% case representing hundreds of billions of operations. Alternatively are you suggesting that the use of an autoloader could lead to incorrect results in real world conditions?

**Reply by Seldaek** [reply]
@markrandall the problem for me is that because we need to support this 1% of cases successfully (and it may well be more than 1%.. but that's beyond the point), I fear we can not use a classmap implementation that does not allow adding/removing a classmap.

> the majority of modern applications are now built with classmaps generated at build time

That is true, classmaps are generated at build time, but there are some use cases where at runtime you end up loading several composer autoloaders, which were all generated at build time, and may all have a classmap. 

One such example is WordPress where many plugins ship their own composer vendor dir, and each plugin when initializing will add its own autoloader from its own vendor dir. Now is this a poor usage pattern? Yes, but it's also reality.

Another example as @stof mentioned is when using PHPUnit or some other tool which was installed separately from the project's vendor dir, either in another directory or because it is shipped as a phar which contains its own composer-autoloader, we end up often having two class loaders present at runtime (the PHPUnit/tool one, and the user project one).

So as I wrote on internals if the PHP API allowed us to append/prepend a classmap, and then remove a classmap by some identifier (which the "add" function could return), that would be much more helpful than having a single classmap.

With the current API proposal, I think we *may* be able to make it work only for the most-common case if we set a classmap the first time a classloader is registered, and if any other classloader is registered or any classloader is unregistered, we would de-optimize entirely and remove the classmap using `autoload_set_classmap([])`. I am still not so satisfied about this though because it means if some other piece of code appended a classmap to ours, we would not know and would just end up wiping it out. The alternative is to fetch the classmap, diff the contents with the contents we are trying to remove, and then rewrite the rest, but that would most likely cancel out any performance gain. 

I hope this explains better why I proposed the API I proposed, which follows the spl_autoload_register pattern and which I will put up again here for completeness:

```php
    autoload_register_classmap(array $map, bool $prepend = false): string|int; // return value is some unique identifier
    autoload_unregister_classmap(string|int $identifier);
```

**Reply by markrandall** [reply]
> With the current API proposal, I think we _may_ be able to make it work only for the most-common case if we set a classmap the first time a classloader is registered, and if any other classloader is registered or any classloader is unregistered, we would de-optimize entirely and remove the classmap using `autoload_set_classmap([])`. 

@Seldaek Under the RFC in its current form this would be an error as the classmap is only allowed to be set once per request (and throws an exception if one is already set). This can be checked before call by performing `if (empty(autoload_get_classmap())) { .. }`. 

To avoid accidental confusion amongst all readers, I do want to be sure it's understood that the inbuilt classmap is not authoritative - if the class does not exist in its table traditional spl_autoload_register behaviour will immediately follow it. 

Let me enquire this of you, what benefit would you forsee in "de-optimizing entirely" if a second loader is set? Currently there is no mechanism to detect if an existing additional autoloader has been added at any point short of repeatedly iterating over spl_autoload_functions and checking each one after any external call which might have added to the list?

I might be convinced that allowing multiple classmaps is a worthwhile idea, but prepending and unregistering them? What's the use case? Can you provide real-world cases for me to look to help me understand where prepending an array of classmaps would yield a different result, or worse performance? (I'm trying to keep the surface as small as possible).

If it makes sense to do so, I will add it, but in so far as I can tell, Composer itself does not have a way of removing its own internal classmaps because the addClassMap function performs a merge if one is already defined. Is this correct?

**Reply by Seldaek** [reply]
> Under the RFC in its current form this would be an error as the classmap is only allowed to be set once per request (and throws an exception if one is already set). This can be checked before call by performing if (empty(autoload_get_classmap())) { .. }.

I missed that fact, but that only complicates matters IMO as it doesn't even allow de-optimization (more on this below).

> To avoid accidental confusion amongst all readers, I do want to be sure it's understood that the inbuilt classmap is not authoritative - if the class does not exist in its table traditional spl_autoload_register behaviour will immediately follow it.

That is my understanding yes, and that's fine. Composer's classmap works much the same way.

> Let me enquire this of you, what benefit would you forsee in "de-optimizing entirely" if a second loader is set? Currently there is no mechanism to detect if an existing additional autoloader has been added at any point short of repeatedly iterating over spl_autoload_functions and checking each one after any external call which might have added to the list?
>
> I might be convinced that allowing multiple classmaps is a worthwhile idea, but prepending and unregistering them? What's the use case? Can you provide real-world cases for me to look to help me understand where prepending an array of classmaps would yield a different result, or worse performance? (I'm trying to keep the surface as small as possible).
>
> If it makes sense to do so, I will add it, but in so far as I can tell, Composer itself does not have a way of removing its own internal classmaps because the addClassMap function performs a merge if one is already defined. Is this correct?

Composer does not add/remove classmaps from one single ClassLoader instance, but in the examples I gave you earlier (phpunit or wordpress), multiple ClassLoaders are registered via spl_autoload_register, and each may have a different classmap.

Now the issue is we do allow prepending ClassLoaders, and each has its own classmap, so if the core-classmap functionality does not let us mimic that, we will inevitably run into edge cases where some class is loaded from the wrong location. Let me try to illustrate this with a common example we ran into in the early composer days:

- PHPUnit required symfony/yaml
- A symfony project requires symfony/yaml too
- I run PHPUnit which I installed somewhere globally on my machine and it was installed with symfony/yaml 5.x 
- PHPUnit bootstrapping loads its own ClassLoader, which has a classmap with `Symfony\Component\Yaml\... => '/global/vendor/symfony/yaml/...` (again 5.x)
- Now PHPUnit loads my test bootstrap file, which includes `/var/project/vendor/autoload.php`, and my project runs on Symfony 4.x right now so that contains symfony/yaml 4.x
- The default behavior is that composer classloader are prepended, so the latter one (my project's) wins over the initial classloader (PHPUnit's).
- Once the Yaml class is needed, it gets loaded from `/var/project/vendor/symfony/yaml/...` - which is the most correct behavior as that is what my project depends on. It's not a perfect world as some incompatibilities may arise between project and PHPUnit dependencies, but it's the best we can do really.

If PHPUnit's classloader instead was populating the core classmap with `autoload_set_classmap()` - you'd get PHPUnit's Yaml 5.x loaded instead, which would be a different behavior.

**Reply by markrandall** [reply]
@Seldaek Thank you for the detailed description. I'll have to have a ponder, but that feels distinctly like an unhandled error condition to me, one I don't think we currently have any way of easily detecting in advance short of prepending an autoloader that calls _every_ autoloader and sees if it blows up before getting to the end with a class redefined error.

Multiple classmaps would perhaps offer at least some way around that, as we could offer a debugging capability for checking for collisions between the keys. It certainly wouldn't be fast, and could only be used in dev, but it would at least provide for some insight vs what is currently available.

**Reply by Seldaek** [reply]
I think you misunderstood my scenario above. There is no class being redefined. The only difference between a prepended or appended autoloader is whether the symfony/yaml 4.x or 5.x class gets loaded. Once the class is loaded the spl_autoloader chain is broken and aborted, so really what I would expect to be able to do is:

- when PHPUnit autoloader is registered, call `autoload_register_classmap($this->classmap, $prepend)`
- when project's autoloader is registered, call `autoload_register_classmap($this->classmap, $prepend)`
- when Yaml class is needed, core would look at classmap arrays in order, and the first one having $classname as key gets included and that is the end of it.
- If no match in any classmap, forward to spl_autoloaders

**Reply by markrandall** [reply]
@Seldaek My comment was about the ability to detect conflicts with multiple autoloaders providing different files, the only way I could think of detecting this would be to call every registered autoloader in turn, and if it attempted to redeclare a class, you would know that multiple autoloaders were attempting to provide the same class name and may have a conflict.

In your example, PHPUnit would silently be loaded with a potentially incompatible Yaml class, is that correct?

**Reply by Seldaek** [reply]
> My comment was about the ability to detect conflicts with multiple autoloaders providing different files

IMO detecting autoloader conflict is a userland concern, and not something you should consider when designing this feature. If done by introspecting ClassLoaders one can do this without triggering errors.

> In your example, PHPUnit would silently be loaded with a potentially incompatible Yaml class, is that correct?

Right, that's what I meant with "It's not a perfect world as some incompatibilities may arise between project and PHPUnit dependencies, but it's the best we can do really.".

It's usually better this way than the app being tested being loaded with an incompatible Yaml class. Hence why we default to prepend, but you can switch it to append if you rather. The point is again that if the core classmap thing facilitates that, user are free to do it however they need/want. If it's a single classmap we cannot implement this correctly.

**Reply by stof** [reply]
> It's usually better this way than the app being tested being loaded with an incompatible Yaml class. Hence why we default to prepend, but you can switch it to append if you rather. The point is again that if the core classmap thing facilitates that, user are free to do it however they need/want. If it's a single classmap we cannot implement this correctly.

@Seldaek note that this won't keep working magically. When developing, the app autoloader won't generate a classmap but rely on the PSr-4 discovery. But a PHPUnit phar will likely contain an optimized autoloader. And so we would end up with a case where the PHPUnit YAML is loaded from the PHP classmap because the order of autoloading becomes `phpunit classmap > app ClassLoader > phpunit ClassLoader`

**Reply by Seldaek** [reply]
Ah true, that makes the entire thing quite unreliable for multiple-autoloaders contexts.

As I said above, I would be fine supporting this only if a single ClassLoader is present, but we still need a way to remove the classmap IMO in case we notice a second ClassLoader is being created, in which case we should remove the php-core classmap and go back to "slow" mode.

@markrandall what do you think about that? It'd mean you keep the current API, and we use it where possible, and so we can support the cases where it's not possible to use, you'd need to provide a `autoload_reset_classmap()` or `autoload_clear_classmap()` or similar..

It keeps things simple in core, and allows us to ensure compatibility so we don't end up creating hard to debug issues for people by adding an small optimization.

**Reply by markrandall** [reply]
@Seldaek How would you go about detecting multiple class loaders being brought up?

Clearing the classmap is an easy add, I could remove the restriction on setting it multiple times, allowing it to be set to empty, or another list. Last one to be set wins.

But just to make sure we're all on the same page here, is the current thinking that using multiple sequential classmaps (with optional prepending) would cause more problems than it helps during development due to things like phpunit ending up using their own compiled lists which would supersede any traditional PSR4 autoloaders?

**Reply by Seldaek** [reply]
Yeah, I think either in dev or even in prod, if some classloaders are optimized and some are not, it would cause potential issues, so I'd rather not do it. Then only the sane environments with a single optimized ClassLoader in prod would benefit from the optimization, which is fine by me.

> How would you go about detecting multiple class loaders being brought up?

We already have a way to keep track of multiple loaders for some other feature using https://github.com/composer/composer/blob/master/src/Composer/Autoload/ClassLoader.php#L62 - so I think it should be doable in register() to know if this ClassLoader is not the first one then we disable core-classmap optimizations.

**Reply by stof** [reply]
@Seldaek I would add that *appending* an autoloader should not disable the PHP classmap, only *prepending* a composer autoloader. This way, we avoid disabling the prod optimization if a project creates a second ClassLoader instance for its customs needs and append it after the generated one (which is better than changing dynamically the config of the generated autoloader, which would force a copy of the config out of shared memory and might be incompatible with composer's authoritative mode).
An appended autoloader is not affected by the presence of an existing PHP classmap (assuming it does not try to set it of course), as it does not care which previous items of the autoloading stack are handling classes before it gets called.

**Reply by markrandall** [reply]
@Seldaek Okay I think we're close to a solution then. 

Just so I know how this can be presented to internals, would we also expect this mechanism to be knocked out for the likes of wordpress due to your previous comments where plugins register their own?

**Reply by stof** [reply]
or even better: prepending a composer autoloader could always replace the PHP classmap with its own (as it is meant to run before the autoloaders that were already registered, and the PHP classmap also runs before them). and appending a composer autoloader would only set the classmap if it is the first one being registered, as said in my previous comment (as for the first autoloader, prepending or appending is actually putting it in the same place)

**Reply by stof** [reply]
@markrandall if the various composer plugins ship their own Composer autoloaders (due to being installed in older ways not managing autoloading for them), my proposal would probably cover it already

**Reply by Seldaek** [reply]
@stof right that makes sense 👍🏻 

@markrandall most likely for wordpress this would not optimize much, but tbh the rate of Composer usage in WordPress is anyway unknown, most likely fairly low. Some plugins use it internally but I don't know exactly how wide-spread that is. WP is also mostly deployed without build step, so it doesn't really lend itself well to a classmap autoloader to begin with, and it's IMO not the main target for such a performance optimization.

**Reply by stof** [reply]
@markrandall and yes, for my proposal to work, the only change needed for the RFC is to allow calling it multiple times, with the "last one wins" semantic. Everything else is on the composer side when using it.

**Reply by markrandall** [reply]
@Seldaek @stof If everyone is in agreement, I'm happy to make that change to the RFC.

**Reply by Seldaek** [reply]
Sounds good! Thanks both for the discussion here, hopefully this leads to some meaningful improvement down the line :)
